File: backend/app/models/company.py
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import DateTime, ForeignKey, Index, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base_class import Base


class Company(Base):
    __tablename__ = "companies"
    __table_args__ = (Index("ix_companies_organization_id", "organization_id"),)

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    organization_id: Mapped[UUID] = mapped_column(
        ForeignKey("organizations.id", ondelete="RESTRICT"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    website: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    industry: Mapped[str | None] = mapped_column(String(255), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    organization: Mapped["Organization"] = relationship(back_populates="companies")
    contacts: Mapped[list["Contact"]] = relationship(
        back_populates="company", cascade="all, delete-orphan"
    )
    leads: Mapped[list["Lead"]] = relationship(back_populates="company")
    deals: Mapped[list["Deal"]] = relationship(back_populates="company")
    activities: Mapped[list["Activity"]] = relationship(back_populates="company")
    notes: Mapped[list["Note"]] = relationship(back_populates="company")
    tags: Mapped[list["Tag"]] = relationship(
        secondary="company_tags", back_populates="companies"
    )

File: frontend/src/types/company.ts
export type Company = {
  id: string;
  organization_id: string;
  name: string;
  website: string | null;
  industry: string | null;
  created_at: string;
  updated_at: string;
};

export type CompanyInput = {
  name: string;
  website: string | null;
  industry: string | null;
};

export type CompaniesListParams = {
  industry?: string;
  page?: number;
  page_size?: number;
  search?: string;
  sort_by?: "created_at" | "industry" | "name" | "updated_at";
  sort_direction?: "asc" | "desc";
};
